component/file_cache/file_cache.go

```diff
@@ -1268,11 +1380,28 @@ func (fc *FileCache) TruncateFile(options internal.TruncateFileOptions) error {
 	} else {
 		// If size is not 0 then we need to open the file and then truncate it
 		// Open will force download if file was not present in local system
-		h, err = fc.OpenFile(internal.OpenFileOptions{Name: options.Name, Flags: os.O_RDWR, Mode: fc.defaultPermission})
-		if err != nil {
-			log.Err("FileCache::TruncateFile : Error opening file %s [%s]", options.Name, err.Error())
+		localPath := common.JoinUnixFilepath(fc.tmpPath, options.Name)
+		flock := fc.fileLocks.Get(options.Name)
+		downloadRequired, _, _, err := fc.isDownloadRequired(localPath, options.Name, flock)
+		// return err in case of authorization permission mismatch
+		if err != nil && err == syscall.EACCES {
 			return err
 		}
+		h = handlemap.NewHandle(options.Name)
+		if downloadRequired {
+			_, err = fc.DownloadFile(internal.DownloadFileOptions{Name: options.Name, Flags: os.O_RDWR, Mode: fc.defaultPermission, Handle: h})
+			if err != nil {
+				log.Err("FileCache::TruncateFile : Error opening file %s [%s]", options.Name, err.Error())
+				return err
+			}
+		} else {
+			h, err = fc.OpenFile(internal.OpenFileOptions{Name: options.Name, Flags: os.O_RDWR, Mode: fc.defaultPermission})
+			if err != nil {
+				log.Err("FileCache::TruncateFile : Error opening file %s [%s]", options.Name, err.Error())
+				return err
+			}
+		}
+
```

Comment:
Shouldn't this code just call `fc.OpenFile`, followed by the same code you use in `ReadInBuffer` and `Write` to call `getHandleData`?

Reply:
Okay, I guess this comment isn't about concurrency. But the next one is!